src/monodroid/jni/logger.cc

```diff
@@ -113,93 +114,149 @@ init_reference_logging (const char *override_dir)
 
 	if ((log_categories & LOG_LREF) != 0 && !light_lref) {
 		// if both lref & gref have files specified, and they're the same path, reuse the FILE*.
-		if (lref_file != NULL && strcmp (lref_file, gref_file != NULL ? gref_file : "") == 0) {
+		if (lref_file != nullptr && strcmp (lref_file, gref_file != nullptr ? gref_file : "") == 0) {
 			lref_log  = gref_log;
 		} else {
 			lref_log  = open_file (LOG_LREF, lref_file, override_dir, "lrefs.txt");
 		}
 	}
 }
 
+template<size_t NameSize>
+force_inline static bool
+set_category (const char (&name)[NameSize], string_segment& arg, unsigned int entry, bool arg_starts_with_name = false)
+{
+	if ((log_categories & entry) == entry) {
+		return false;
```

Comment:
Not sure I understand this rationale: I think the `if` is "if this category is already being logged".  If the category *is* already being logged, shouldn't this return `true` so that we hit the `continue` in `init_logging_categories()`?

Consider:

```cpp
set_category ("default",  param, LOG_DEFAULT);
set_category ("default",  param, LOG_DEFAULT);
```

Shouldn't the 2nd invocation have the same return value as the first invocation?

Reply:
No, we can't do that. Consider that each category name is checked against each token in the property value being parsed. If we return `true` from here, then the category names that follow the call in the code won't be matched. Consider the value of `debug.mono.log` being `assembly,default,debugger` and this code:

```cpp
while (value.next_token (',', param)) {
  if (set_category ("assembly", param, LOG_ASSEMBLY)) {
    continue;
  }

  if (set_category ("default",  param, LOG_DEFAULT)) {
    continue;
  }

  if (set_category ("debugger", param, LOG_DEBUGGER))  {
    continue;
  }
```

`assembly` would be matched for the first token and neither `default` nor `debugger` would ever be matched if we returned `true` on line 130. In other words, the return value from `set_category` should be read as `found a match` instead of `category already set`. It's just a tiny optimization.